### DISCUSSION TITLE
Moving ESP32-based Listener Clients' Codebase to Platform.io

### DISCUSSION THREAD

**Discussion by mattv8** [josephdadams/TallyArbiter#269] [Ideas]
I'd like to open a discussion about porting the ESP32 and Arduino C++ based listener clients to PlatoformIO. Simply put, PlatformIO is a robust and **much** more coder-friendly IDE than the default Arduino IDE. For further reading about PlatoformIO, see here: [https://platformio.org/](https://platformio.org/).

I recently "ported" [my listener client's](https://github.com/mattv8/open-tally-wifi-client) code base to PlatoformIO and I've been very happy with the results, although it is still very much a work-in-progress.

Applicable listener clients currently in the TallyArbiter repo:
- M5AtomMatrix-listener
- TTGO_T-listener
- esp32-neopixel-listener
- m5stickc-listene

Pros:
- Support for two different IDE's: Visual Studio Code and Atom.io.
  - Programming therefore becomes much easier by using a full-featured IDE.
- VS Code's[ intellisense](https://code.visualstudio.com/docs/editor/intellisense) would become available for use.
- Faster and better compiler.
- Much better library management.
- Porting from Arduino IDE to Platform.io is a cinch, and I could make these initial contributions.

Cons:
- It would make coding in Arduino's IDE no longer an option.
- It might therefore make contributing to the listener clients codebases less accessible to "novice" programmers who are unfamiliar with PlatoformIO.
  - I feel this could be fixed with good documentation.

Neutrals:
- It would change the directory structure of each respective listener client's codebase.

**Comment by josephdadams** [comment]
I think we are working toward distributing compiled firmware hex files which will not require users to delve much into the Arduino IDE anyways. Isn't that the plan, @hrueger ?

Beyond that though, I'm all for using anything but the Arduino IDE. Gosh it's awful.

**Reply by mattv8** [reply]
Great! If we're distributing compiled firmwares to the end-users anyways, PlatformIO has a pretty good pipeline for creating the firmware binaries.

**Reply by josephdadams** [reply]
GH Actions can't do this? @MatteoGheza has them doing something currently - I just don't know what 😆

**Reply by hrueger** [reply]
Yes, as far as I know, that was the plan.
For me it would be ok if we require users to connect their devices via USB, then we would not have to deal with OTA.

**Reply by josephdadams** [reply]
Yes, I'm also good with USB. Even if USB is the goal now, and in another major release we offer OTA capability.

**Comment by JTF4** [comment]
I would be all for that. I would have to relearn a few things but I think that in the long run it will be better. 

Only question I have is if it's more or less processing intensive than arduino.

Oh....... and I absolutely DESPISE the arduino ide. I actually edit all my arduino stuff in VS code and then only use the arduino thing for compiling.

**Comment by MatteoGheza** [comment]
Working on a new "generic embedded listener client", that works for M5Atom/M5Stick/(generic) ESP32/(generic) ESP8266/TTGO.
Works with PlatformIO envs, so every device firmware is based on the same codebase and is compiled only with required libs/code.
Doing this, fixing bug, supporting new devices and adding new features will be easier.

**Reply by josephdadams** [reply]
This sounds awesome. Thank you!
